src/ports/postgres/modules/sample/balance_sample.py_in

```diff
@@ -0,0 +1,994 @@
+                    SELECT
+                            classes,
+                            CASE (SELECT 1 WHERE desired_values is NULL)
+                                WHEN 1 THEN extra_values
+                                ELSE desired_values
+                            END AS desired_values,
+                            class_count
+                    FROM
+                        {class_counts}
+                        LEFT JOIN
+                        {desired_counts}
+                        ON ({class_counts}.classes = {desired_counts}.desired_classes)
+                        LEFT JOIN
+                        {remaining_class_samples}
+                        ON ({class_counts}.classes = {remaining_class_samples}.extra_classes)
+                        ) AS foo
+                )
+            """.format(**locals())
+        else:
+            """
+                If output_table_size is not specified,
+                return remaining classes as is.
+            """
+            desired_sample_per_class_sql = """
+                CREATE VIEW {desired_sample_per_class} AS (
+                    SELECT
+                            classes,
+                            desired_values::int AS sample_class_size,
+                            class_count as actual_class_size,
+                            CASE
+                              WHEN (desired_values - class_count) = 0 THEN 'nosample'
+                              WHEN (desired_values - class_count) > 0 THEN 'oversample' ELSE 'undersample'
+                            END AS category
+                    FROM (
+                        SELECT
+                                classes,
+                                CASE (SELECT 1 WHERE desired_values is NULL)
+                                    WHEN 1 THEN class_count
+                                    ELSE desired_values
+                                END AS desired_values,
+                                class_count
+                        FROM
+                            {class_counts}
+                             LEFT JOIN
+                            {desired_counts}
+                            ON ({class_counts}.classes = {desired_counts}.desired_classes)
+                            ) AS foo
+                    )
+                """.format(**locals())
+        if remaining_class_samples:
+            temp_views.append(remaining_class_samples)
+
+    plpy.execute(desired_sample_per_class_sql)
+    temp_views.append(desired_sample_per_class)
+    return temp_views
+
+"""
+    Sample the remaining classes in output_table_size - (total calculated class sizes).
+"""
+def _sample_remaining_classes_uniformly_per_output_table_size(class_counts
+                , desired_counts, output_table_size):
+
+    remaining_class_samples = unique_string(desp='remaining_class_samples')
+    foo_table = unique_string(desp='foo')
+    remaining_class_samples_sql = """
+        CREATE VIEW {remaining_class_samples} as (
```

Comment:
Maybe we should read these counts and handle simple computations in the python layer instead of creating a view. It would be easier to read IMHO.

Reply:
Changing this would result in changing many more queries. So leaving this as is for now. We can revisit this if performance is bad.